### PR TITLE
Add parameter flag to loss compialation

### PR DESCRIPTION
**Pull request by ggustavs**
Currently, there is no support for parameters in specifications when compiling to loss. This PR just adds them in the same way as for verification.

### CODE REVIEW

**Comment by MatthewDaggitt** [comment]
Hmm there is supposed to be support already! Parameters and datasets are added as arguments to the generated loss function, e.g.
```
@parameter r : Real

@network f : Real -> Real

@property
p : Bool
p = f r > 0
```
you should then be able to do: 
```python
loss_fn = loss_pt.load_specification("spec.vcl")["p"]
loss = loss_fn(r = 0.3, f = model)
```
Does this not currently work?

**Comment by ggustavs** [comment]
> Hmm there is supposed to be support already! Parameters and datasets are added as arguments to the generated loss function, e.g.
> 
> ```
> @parameter r : Real
> 
> @network f : Real -> Real
> 
> @property
> p : Bool
> p = f r > 0
> ```
> 
> you should then be able to do:
> 
> ```python
> loss_fn = loss_pt.load_specification("spec.vcl")["p"]
> loss = loss_fn(r = 0.3, f = model)
> ```
> 
> Does this not currently work?

Nope:
```shell
uvx --from vehicle-lang vehicle --json compile loss -s pk.vcl -l DL2Loss
Installed 19 packages in 60ms
Something went wrong internally. Please report the error shown below to `https://github.com/vehicle-lang/vehicle/issues`.
Error: encountered unexpected expression:
  non-network free var
during unblocking.
Stack:
CallStack (from HasCallStack):
  developerError, called at src/Vehicle/Prelude/Error.hs:49:33 in vehicle-0.24.0-inplace:Vehicle.Prelude.Error
  unexpectedExprError, called at src/Vehicle/Compile/Unblock.hs:198:12 in vehicle-0.24.0-inplace:Vehicle.Compile.Unblock
```

This will happen for any spec with a parameter as far as my testing has shown and as far as my understanding of the compiler code goes.

**Comment by ggustavs** [comment]
I did not want to make a new branch to fix it, but I this PR also fixes an issue where properly bounded variables were incorrectly being labeled as unbounded if the bounds were explicitly specified for each index of the variable, and not with forall.

**Comment by MatthewDaggitt** [comment]
> This will happen for any spec with a parameter as far as my testing has shown and as far as my understanding of the compiler code goes.

Hmm, so this was supposed to work. I think my proposed approach is the design we're after as (i) it preserves the invariant that all external resources (networks, parameters, datasets) are provided by the user at runtime and (ii) it allows users can calculate the gradient with respect to the parameters as well if they wish. I will try and get an alternative fix up at running.

> I did not want to make a new branch to fix it, but I this PR also fixes an issue where properly bounded variables were incorrectly being labeled as unbounded if the bounds were explicitly specified for each index of the variable, and not with forall.

Thanks. I have looked at this, and there's a function that was already supposed to do the reduction from `x ! 0` to `x0` so I have put an alternative fix up at #1086.

**Comment by MatthewDaggitt** [comment]
Hmm so I'm quite confused. When I run the command on my example above I get the JSON which represents the below as expected. 
```
@property;
p : Tensor Real -> (Tensor Real -> Tensor Real) -> Tensor Real;
p = \ r -> \ f -> max 0.0 (0.0 - f r)
```

This works on both `0.23` and on `dev` branch.

Could you give me a folder with the exact files + command that you're running as well as your vehicle version?

**Comment by ggustavs** [comment]
> This works on both `0.23` and on `dev` branch.
> 
> Could you give me a folder with the exact files + command that you're running as well as your vehicle version?

Weeeeird

```
❯ uvx --from vehicle-lang vehicle compile loss -s test_spec.vcl -l DL2Loss
Something went wrong internally. Please report the error shown below to `https://github.com/vehicle-lang/vehicle/issues`.
Error: encountered unexpected expression:
  non-network free var
during unblocking.
Stack:
CallStack (from HasCallStack):
  developerError, called at src/Vehicle/Prelude/Error.hs:49:33 in vehicle-0.24.0-inplace:Vehicle.Prelude.Error
  unexpectedExprError, called at src/Vehicle/Compile/Unblock.hs:198:12 in vehicle-0.24.0-inplace:Vehicle.Compile.Unblock

cloned_vehicle on  dev [?] via λ 9.4.8
❯ uvx --from vehicle-lang vehicle --version
0.24.0

cloned_vehicle on  dev [?] via λ 9.4.8
❯ cabal run exe:vehicle -- compile loss -s test_spec.vcl -l DL2Loss
Something went wrong internally. Please report the error shown below to `https://github.com/vehicle-lang/vehicle/issues`.
Error: encountered unexpected expression:
  non-network free var
during unblocking.
Stack:
CallStack (from HasCallStack):
  developerError, called at src/Vehicle/Prelude/Error.hs:49:33 in vehicle-0.24.0-inplace:Vehicle.Prelude.Error
  unexpectedExprError, called at src/Vehicle/Compile/Unblock.hs:197:12 in vehicle-0.24.0-inplace:Vehicle.Compile.Unblock

cloned_vehicle on  dev [?] via λ 9.4.8
❯ cabal run exe:vehicle -- --version
0.24.0+80a54f5.dirty
```

[test_spec.vcl](https://github.com/user-attachments/files/25965584/test_spec.vcl.txt)

I also tried deleting my ~/.vehicle and rebuilding on the dev branch, running with 0.23.0 from PyPI - either way, same result.

Buuut, when I try running any of the above configurations with your example, it works??? Im even more confused now :D

**Comment by ggustavs** [comment]
Closed this PR, because I opened [pr #1090](https://github.com/vehicle-lang/vehicle/pull/1090) that fixes [issue #1089 ](https://github.com/vehicle-lang/vehicle/issues/1089)
